### PR TITLE
Module5

### PR DESCRIPTION
**Pull request by Yb1ku**
El trabajo de este módulo consiste en analizar el comportamiento de un modelo LightGBM entrenado para predecir si una acción superará al S&P 500. En los resultados de train y test se observa que hay una diferencia significativa entre ambos grupos, y el test no obtiene buenos resultados. He realizado un análisis con SHAP para ver la influencia de las variables y he visto que el modelo está utilizando casi exclusivamente las variables `s500-change_730`, `close_sp500_0`, `close_0`, `std_minus_730` y `std_minus_120`, y no está usando información de los quarter reports de las empresas. Creo que puede estar relacionado con la gran cantidad de valores nulos que hay en las variables sacadas de los quarter reports, pero tengo entendido que LightGBM maneja internamente el tema de los valores nulos. Además, el modelo está encontrando un patrón claro entre valores bajos de `close_0` y probabilidades altas de que la acción supere al índice, y no es buena idea que el modelo solo se fije en eso (a la vista están los resultados en el test set). Mi hipótesis es que está pasando una de estas dos cosas: 

1. Hay algún tipo de problema con los datos (escala, leakage, etc.) que no he sabido identificar. 
2. Es un problema que tiene que ver con el mundo financiero, algún tipo de mecanismo que utilizan las empresas para bajar artificialmente su cotización ante perspectivas de crecimiento.

### CODE REVIEW

**Review by gscharly — APPROVED** [review]
Hola Diego,

Perdona que haya tardado tanto en revisar el módulo, estuve fuera la semana pasada.

- Me gusta la primera gráfica donde muestras la diferencia entre error de validación y entrenamiento, pero puede tener un poco de trampa. Así como te muestra que hay menos diferencia entre ambas, no te muestra la magnitud de la métrica. Entrenamiento y validación pueden tener valores muy similares, pero también muy bajos. Puedes complementar esa gráfica con uno de los dos valores.

- No conocía la validación adversaria, me parece un método interesante. Si lo estoy interpretando bien, estás consiguiendo un AUC=1 para todas las fechas, lo que significa que el modelo que entrenas con entrenamiento y validación reconoce perfectamente ambas clases? Me parece un valor de AUC bastante elevado, y también me parece raro que siempre sea el mismo (teniendo en cuenta que usas unos hiperparámetros "aleatorios"). Es posible que estés usando un único modelo para todas las fechas?

- Creas una gráfica interesante al plotear las curvas de pérdida en función del número de árboles. Para este modelo en concreto, se ve que a partir de cierto número de árboles, la pérdida de test se estanca. Si repites esto para otras fechas, es posible que veas el mismo efecto, y esto puede ayudar a reducir tiempo y coste de entrenamiento.

- Buen análisis de importancia de variables, aunque de nuevo, fijarse solo en uno de los modelos puede inducir a error. Podrías hacer un análisis similar combinando las importancias de todos los modelos. Si ves variables que a priori parecen extrañas, como close_0, siempre ayuda mirar ejemplos concretos. En la fecha donde ese variable tiene mayor importancia, hay alguna empresa donde pase esto? Puede algún efecto financiero explicarlo? Puede incluir algún tipo de leakage?

- Buena observación con los valores nulos. Aquí, te recomiendo que tengas "cuidado" con implementaciones que tratan valores nulos por defecto. Creo que lightgbm sustituye valores nulos por una categoría NA, lo que puede ser o no beneficioso según que casos. Yo siempre recomiendo tratar los nulos de antemano.
